v-next/core/src/internal/tasks/task-manager.ts

```diff
@@ -36,82 +42,90 @@ export class TaskManagerImplementation implements TaskManager {
       }
     }
 
+    // reduce global user defined tasks
     for (const taskDefinition of this.#hre.config.tasks) {
       this.#reduceTaskDefinition(globalParameterIndex, taskDefinition);
     }
   }
 
   public getTask(taskId: string | string[]): Task {
     taskId = Array.isArray(taskId) ? taskId : [taskId];
-
     if (taskId.length === 0) {
-      throw new Error(`Invalid task id "${formatTaskId(taskId)}"`);
+      throw new HardhatError(
+        HardhatError.ERRORS.TASK_DEFINITIONS.TASK_NOT_FOUND,
+        {
+          task: formatTaskId(taskId),
+        },
+      );
     }
 
     let tasks = this.#rootTasks;
-    let task: Task;
-
+    let task: Task | undefined;
     for (let i = 0; i < taskId.length; i++) {
       const idFragment = taskId[i];
-
       const currentTask = tasks.get(idFragment);
-
       if (currentTask === undefined) {
-        throw new Error(
-          `Task "${formatTaskId(taskId.slice(0, i + 1))}" not found.`,
+        throw new HardhatError(
+          HardhatError.ERRORS.TASK_DEFINITIONS.TASK_NOT_FOUND,
+          {
+            task: formatTaskId(taskId.slice(0, i + 1)),
+          },
         );
       }
 
       task = currentTask;
       tasks = task.subtasks;
     }
 
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- Can't be undefined as we set it in the non-empty loop.
-    return task!;
+    // task can't be undefined as we set it in the non-empty loop
+    assertHardhatInvariant(task !== undefined, "Task not found");
```

Comment:
This comment should be part of the assertion, so that it's self-descriptive, and more helpful it it ever fails:

```suggestion
    assertHardhatInvariant(task !== undefined, "Task is undefined despite it being always set by a non-empty loop");
```